# Building a database for Crime Reports

In this project, we will build a database for storing data related with crimes that occurred in Boston. We'll be using Postgres.

### Goals

The goal of this guided project is for you to create a database named `crimes_db` with a table – `boston_crimes` – with appropriate datatypes for storing the data from the `boston.csv` file. 

We will be creating the table inside a schema named `crimes`. 

We will also create the `readonly` and `readwrite` groups with the appropriate privileges as we have discussed in this course. Finally, you will also need to create one user for each of these groups.

<img src="goal.png">

### Creating Crimes Database

Let's start by creating the `crimes_db` database. We'll first connect to the `dq` database with user `dq` when connecting to the postgres server. We'll create our database, then connect to and setup the scheme - `crimes`

In [66]:
import psycopg2

from psycopg2.extensions import AsIs

In [2]:
# Connect to postgres database
conn = psycopg2.connect(dbname = "postgres", user = "postgres")
conn.autocommit = True
cur = conn.cursor()

In [3]:
# Create crimes database
cur.execute("CREATE DATABASE crimes_db;")
conn.autocommit = False

DuplicateDatabase: database "crimes_db" already exists


In [4]:
conn.close()

In [5]:
# Connect to our newley created database
conn = psycopg2.connect(dbname = "crimes_db", user = "postgres")
cur = conn.cursor()

In [6]:
# Create a schema within our database

cur.execute("CREATE SCHEMA crimes;")

DuplicateSchema: schema "crimes" already exists


### Obtaining Column names and sample

Let's take a look at our Boston data

In [7]:
import csv

In [8]:
# Print header and first two rows of our data
with open("Boston.csv", "r") as file:
    reader = csv.reader(file)
    col_headers = next(reader)
    first_row = next(reader)
    print(col_headers)
    print(first_row)

['incident_number', 'offense_code', 'description', 'date', 'day_of_the_week', 'lat', 'long']
['1', '619', 'LARCENY ALL OTHERS', '2018-09-02', 'Sunday', '42.35779134', '-71.13937053']


### Creating an Auxillary Function

Let's create a function that will tell us all the different values for each column in our data

In [9]:
def get_col_set(csv_filename, col_index):
    values = set()
    with open(csv_filename, 'r') as file:
        next(file)
        reader = csv.reader(file)
        for row in reader:
            values.add(row[col_index])
    return values

In [10]:
print(len(col_headers))

7


In [11]:
for index in range(len(col_headers)):
    values = get_col_set("Boston.csv", index)
    print(f'{col_headers[index]:20} {len(values)}')

incident_number      298329
offense_code         219
description          239
date                 1177
day_of_the_week      7
lat                  18177
long                 18177


### Finding the maximum length

Let's compute the maximum length of each value in the `description` column and `offense_code` column.

In [12]:
description_values = get_col_set("Boston.csv", 2)

max_len = 0
for description in description_values:
    if len(description) > max_len:
        max_len = len(description)
print(max_len)

58


In [13]:
offense_values = get_col_set("Boston.csv", 1)

max_len = 0
for code in offense_values:
    if len(code) > max_len:
        max_len = len(code)
print(max_len)

4


### Creating the Table

Let's now create a table to store the Boston crime data. We can see that for `day_of_the_week` there are 7 unique values. For this, we might want to use an `enumerate datatype`. Let's begin.

In [14]:
print(get_col_set("Boston.csv", 4))

{'Tuesday', 'Saturday', 'Sunday', 'Friday', 'Monday', 'Thursday', 'Wednesday'}


In [17]:
cur.execute("""CREATE TYPE days AS ENUM 
                ('Saturday', 'Wednesday', 'Thursday', 'Sunday', 'Friday', 'Monday', 'Tuesday');""")

DuplicateObject: type "days" already exists


In [18]:
print(col_headers)

['incident_number', 'offense_code', 'description', 'date', 'day_of_the_week', 'lat', 'long']


In [19]:
print(first_row)

['1', '619', 'LARCENY ALL OTHERS', '2018-09-02', 'Sunday', '42.35779134', '-71.13937053']


For these remaining columns, what datatypes should we use?

* `incident_number` - Because numbers go up to 298329, we'll use the `integer` datatype


* `offense_code` - We'll also go with the `smallint` datatype here


* `description` - As this is text, and the maximum length is 58. We'll round up and use the `varchar` datatype with a value of 60.


* `date` - We'll use the `date` type here, which will allow us to perform sane queries.


* `day_of_the_week` - We'll go with our `enumerated data` type. Why? Because because we have a predefined set of values. When we create an enumerated datatype, each possible value is assigned to a 4 byte index within a seperate table. This will ultimately save us memory.


* `lat` - Because we're dealing with decimals, we'll go with the `decimal` datatype


* `long` - Same as above

In [20]:
# Create our boston_crimes table within our new schema
try:
    cur.execute("""CREATE TABLE crimes.boston_crimes (
                incident_number integer PRIMARY KEY,
                 offense_code smallint,
                 description varchar(60),
                 date date,
                 day_of_the_week days,
                 lat decimal,
                 long decimal);""")
except Exception as ex:
    print(ex)

current transaction is aborted, commands ignored until end of transaction block



In [24]:
cur.execute("SELECT * FROM crimes.boston_crimes LIMIT 10;")

In [25]:
for i in cur.description:
    print(i)

Column(name='incident_number', type_code=23)
Column(name='offense_code', type_code=21)
Column(name='description', type_code=1043)
Column(name='date', type_code=1082)
Column(name='day_of_the_week', type_code=16468)
Column(name='lat', type_code=1700)
Column(name='long', type_code=1700)


In [26]:
conn.commit()
conn.close()

### Loading the Data

Let's now copy our Boston data into the database. We'll use the `copy_expert` method as below.

In [28]:
# Connect again to our database
conn = psycopg2.connect(dbname = "crimes_db", user = "postgres")
cur = conn.cursor()

In [29]:
with open("Boston.csv", "r") as f:
    cur.copy_expert("COPY crimes.boston_crimes FROM STDIN WITH CSV HEADER;", f)

UniqueViolation: duplicate key value violates unique constraint "boston_crimes_pkey"
DETAIL:  Key (incident_number)=(1) already exists.
CONTEXT:  COPY boston_crimes, line 2


In [32]:
cur.execute("SELECT * FROM pg_user;")

In [33]:
cur.fetchall()

[('postgres', 10, True, True, True, True, '********', None, None),
 ('aaronwright', 16384, True, True, False, False, '********', None, None),
 ('data_analyst', 16493, False, False, False, False, '********', None, None),
 ('data_scientist', 16494, False, False, False, False, '********', None, None)]

### Revoking Public Privileges

When we created our database `crimes_db` a `public` schema was automatically created, and grants access to this schema via a role/group `public`. 

We want to revoke all privliges from the `public` group on the `public` schema.

We also want to revoke all privliges of `public` on the `crimes_db` database.

If we don't do this, a user will inherit privileges from the `public` role even if we restrict their privileges and will be able to do stuff with the `public` schema (e.g. create objects)

In [34]:
cur.execute("REVOKE ALL ON SCHEMA public FROM public;")

In [35]:
cur.execute("REVOKE ALL ON DATABASE crimes_db FROM public;")

### Creating User Groups

Let's now setup some groups. Mainly we want `readonly` and `readwrite` groups setup.

In [36]:
# Create our groups/roles
cur.execute("CREATE GROUP readonly NOLOGIN;")
cur.execute("CREATE GROUP readwrite NOLOGIN;")

DuplicateObject: role "readonly" already exists


In [39]:
# Grant our groups/roles connections prviliges to our database
cur.execute("GRANT CONNECT ON DATABASE crimes_db TO readonly;")
cur.execute("GRANT CONNECT ON DATABASE crimes_db TO readwrite;")

In [40]:
# Grant usage privilages to our schema
cur.execute("GRANT USAGE ON SCHEMA crimes TO readonly;")
cur.execute("GRANT USAGE ON SCHEMA crimes TO readwrite")

In [41]:
# Grant prvilieges to our groups
cur.execute("GRANT SELECT ON ALL TABLES IN SCHEMA CRIMES TO readonly;")
cur.execute("GRANT SELECT, INSERT, DELETE, UPDATE ON ALL TABLES IN SCHEMA CRIMES TO readwrite;")

In [42]:
# Check our privileges are correct
cur.execute("SELECT grantor, grantee, privilege_type FROM information_schema.table_privileges WHERE table_name = 'boston_crimes';")

In [43]:
cur.fetchall()

[('postgres', 'postgres', 'INSERT'),
 ('postgres', 'postgres', 'SELECT'),
 ('postgres', 'postgres', 'UPDATE'),
 ('postgres', 'postgres', 'DELETE'),
 ('postgres', 'postgres', 'TRUNCATE'),
 ('postgres', 'postgres', 'REFERENCES'),
 ('postgres', 'postgres', 'TRIGGER'),
 ('postgres', 'readonly', 'SELECT'),
 ('postgres', 'readwrite', 'INSERT'),
 ('postgres', 'readwrite', 'SELECT'),
 ('postgres', 'readwrite', 'UPDATE'),
 ('postgres', 'readwrite', 'DELETE')]

### Creating Users

Let's now create one user in each group, giving them a password.

In [44]:
# Set up users
cur.execute("CREATE USER data_analyst WITH PASSWORD 'secret1';")
cur.execute("CREATE USER data_scientist WITH PASSWORD 'secret2';")

DuplicateObject: role "data_analyst" already exists


In [45]:
# Assign users to groups
cur.execute("GRANT readonly TO data_analyst;")
cur.execute("GRANT readwrite TO data_scientist;")

InFailedSqlTransaction: current transaction is aborted, commands ignored until end of transaction block


In [47]:
conn.commit()
conn.close()

### Testing

Let's now test we've set everything up correctly

In [55]:
conn = psycopg2.connect(dbname = "crimes_db", user = "postgres")

In [56]:
cur = conn.cursor()

In [50]:
# Check our roles have the right prvileges
cur.execute("""SELECT grantee, privilege_type
                FROM information_schema.table_privileges
                WHERE grantee = 'readwrite' OR grantee = 'readonly';""")
cur.fetchall()

[('readonly', 'SELECT'),
 ('readwrite', 'INSERT'),
 ('readwrite', 'SELECT'),
 ('readwrite', 'UPDATE'),
 ('readwrite', 'DELETE')]

In [51]:
# Check all of our existing tables by querying existing tables contained in Postgres engine
cur.execute("SELECT tablename FROM pg_catalog.pg_tables ORDER BY 1;")
cur.fetchall()

[('boston_crimes',),
 ('pg_aggregate',),
 ('pg_am',),
 ('pg_amop',),
 ('pg_amproc',),
 ('pg_attrdef',),
 ('pg_attribute',),
 ('pg_auth_members',),
 ('pg_authid',),
 ('pg_cast',),
 ('pg_class',),
 ('pg_collation',),
 ('pg_constraint',),
 ('pg_conversion',),
 ('pg_database',),
 ('pg_db_role_setting',),
 ('pg_default_acl',),
 ('pg_depend',),
 ('pg_description',),
 ('pg_enum',),
 ('pg_event_trigger',),
 ('pg_extension',),
 ('pg_foreign_data_wrapper',),
 ('pg_foreign_server',),
 ('pg_foreign_table',),
 ('pg_index',),
 ('pg_inherits',),
 ('pg_init_privs',),
 ('pg_language',),
 ('pg_largeobject',),
 ('pg_largeobject_metadata',),
 ('pg_namespace',),
 ('pg_opclass',),
 ('pg_operator',),
 ('pg_opfamily',),
 ('pg_partitioned_table',),
 ('pg_policy',),
 ('pg_proc',),
 ('pg_publication',),
 ('pg_publication_rel',),
 ('pg_range',),
 ('pg_replication_origin',),
 ('pg_rewrite',),
 ('pg_seclabel',),
 ('pg_sequence',),
 ('pg_shdepend',),
 ('pg_shdescription',),
 ('pg_shseclabel',),
 ('pg_statistic',),
 

Here we can see a lot of tables starting with **pg**. These are part of the `pg_catalog` group of tables built into the Postgres engine. So how to we differentiate between tables that are user defined, and those that are inernal?

We can do this by thinking about **schemas**. These are essentially folders for seperating out tables.

When a database is created, 3 schemas are created - `pg_catalog` (for system catalog tables), `information_schema` (for info schema tables), and `public` (the default schemas for creating tables)

When you issue a `CREATE TABLE` command, this will be default be stored under the `public` schema.

So, if we want to see what user created tables have been created, we can filter our above query to exclude tables existing under `pg_catalog` or `information_schema`.

In [52]:
cur.execute("""SELECT tablename, 
                      schemaname 
               FROM pg_catalog.pg_tables 
               WHERE schemaname != 'pg_catalog' AND
                     schemaname != 'information_schema'
               ORDER BY 1;""")
cur.fetchall()

[('boston_crimes', 'crimes')]

In [53]:
cur.execute("SELECT * FROM crimes.boston_crimes LIMIT 10;")
rows = cur.fetchall()
print(type(rows))
for row in rows:
    print(row)

<class 'list'>
(1, 619, 'LARCENY ALL OTHERS', datetime.date(2018, 9, 2), 'Sunday', Decimal('42.35779134'), Decimal('-71.13937053'))
(2, 1402, 'VANDALISM', datetime.date(2018, 8, 21), 'Tuesday', Decimal('42.30682138'), Decimal('-71.06030035'))
(3, 3410, 'TOWED MOTOR VEHICLE', datetime.date(2018, 9, 3), 'Monday', Decimal('42.34658879'), Decimal('-71.07242943'))
(4, 3114, 'INVESTIGATE PROPERTY', datetime.date(2018, 9, 3), 'Monday', Decimal('42.33418175'), Decimal('-71.07866441'))
(5, 3114, 'INVESTIGATE PROPERTY', datetime.date(2018, 9, 3), 'Monday', Decimal('42.27536542'), Decimal('-71.09036101'))
(6, 3820, 'M/V ACCIDENT INVOLVING PEDESTRIAN - INJURY', datetime.date(2018, 9, 3), 'Monday', Decimal('42.29019621'), Decimal('-71.07159012'))
(7, 724, 'AUTO THEFT', datetime.date(2018, 9, 3), 'Monday', Decimal('42.30607218'), Decimal('-71.0827326'))
(8, 3301, 'VERBAL DISPUTE', datetime.date(2018, 9, 3), 'Monday', Decimal('42.32701648'), Decimal('-71.10555088'))
(9, 301, 'ROBBERY - STREET', datet

In [58]:
conn.close()

### Let's play around with this database

In [60]:
conn = psycopg2.connect(dbname = "crimes_db", user = "postgres")
cur = conn.cursor()

Let's first have a look at the `pg_catalog.pg_tables` table.

In [65]:
cur.execute("SELECT tablename FROM pg_catalog.pg_tables WHERE schemaname != 'pg_catalog' AND schemaname != 'information_schema' LIMIT 10;")
table_names = cur.fetchall()

for row in table_names:
    print(row)

('boston_crimes',)


Above we get some information. We have the `boston_crimes` table. We filtered out the `pg_catalog` and `information_schema` schemas, as these are internal tables which contain info we're not interested in right now.

If we wanted to loop through all tables (note we only have one in this instance) and save a description of them down, we can do this:

In [84]:
col_description = dict()
for table in table_names:
    cur.execute("SELECT * FROM crimes.%s LIMIT 0;", (AsIs(table[0]),))
    col_description[table[0]] = cur.description

In [85]:
col_description

{'boston_crimes': (Column(name='incident_number', type_code=23),
  Column(name='offense_code', type_code=21),
  Column(name='description', type_code=1043),
  Column(name='date', type_code=1082),
  Column(name='day_of_the_week', type_code=16468),
  Column(name='lat', type_code=1700),
  Column(name='long', type_code=1700))}

In the above, we can see that in the `boston_crimes` table, we have our columns, along with a type code. What is this type code? This represents the datatype.

In [96]:
cur.execute("SELECT oid, typname FROM pg_catalog.pg_type LIMIT 10;")
cur.fetchall()

[(16, 'bool'),
 (17, 'bytea'),
 (18, 'char'),
 (19, 'name'),
 (20, 'int8'),
 (21, 'int2'),
 (22, 'int2vector'),
 (23, 'int4'),
 (24, 'regproc'),
 (25, 'text')]

We can see what these numbers corrospond to. Let's add all this to a dictionary.

In [109]:
type_code_dict = dict()
cur.execute("SELECT oid, typname FROM pg_catalog.pg_type;")
types = cur.fetchall()

for value in types:
    code = value[0]
    d_type = value[1]
    type_code_dict[code] = d_type
    
list(type_code_dict.items())[:4]

[(16, 'bool'), (17, 'bytea'), (18, 'char'), (19, 'name')]

In [115]:
col_description

{'boston_crimes': (Column(name='incident_number', type_code=23),
  Column(name='offense_code', type_code=21),
  Column(name='description', type_code=1043),
  Column(name='date', type_code=1082),
  Column(name='day_of_the_week', type_code=16468),
  Column(name='lat', type_code=1700),
  Column(name='long', type_code=1700))}

In [144]:
readable_description = {"boston_crimes" : [{"name" : col.name,
                                            "type" : type_code_dict[col.type_code],
                                            "internal size" : col.internal_size} 
                                           for col in col_description["boston_crimes"]]}

In [145]:
for row in readable_description["boston_crimes"]:
    print(row)

{'name': 'incident_number', 'type': 'int4', 'internal size': 4}
{'name': 'offense_code', 'type': 'int2', 'internal size': 2}
{'name': 'description', 'type': 'varchar', 'internal size': 60}
{'name': 'date', 'type': 'date', 'internal size': 4}
{'name': 'day_of_the_week', 'type': 'days', 'internal size': 4}
{'name': 'lat', 'type': 'numeric', 'internal size': -1}
{'name': 'long', 'type': 'numeric', 'internal size': -1}


Above, we've created a new dictionary. We didn't really need to in this instance because we only have one table - `boston_crimes` - but if we had more it would be useful to have a storage place for all the info. The info itself is giving us the name, type and internal size of the columns within our table.

Now, let's add a new key which represents number of rows in the table.

In [146]:
cur.execute("SELECT COUNT(*) FROM crimes.boston_crimes;")
no_of_rows = cur.fetchone()[0]
readable_description["no_of_rows"] = no_of_rows
readable_description

{'boston_crimes': [{'name': 'incident_number',
   'type': 'int4',
   'internal size': 4},
  {'name': 'offense_code', 'type': 'int2', 'internal size': 2},
  {'name': 'description', 'type': 'varchar', 'internal size': 60},
  {'name': 'date', 'type': 'date', 'internal size': 4},
  {'name': 'day_of_the_week', 'type': 'days', 'internal size': 4},
  {'name': 'lat', 'type': 'numeric', 'internal size': -1},
  {'name': 'long', 'type': 'numeric', 'internal size': -1}],
 'no_of_rows': 298329}

In [152]:
conn.commit()
conn.close()

Great. However, this isn't very readable. Let's convert it to JSON format. This is essentialy a way to represent dictionaries in text files.

In Python, we have the `json` library to help with this.

In [153]:
import json
readable_description_json = json.dumps(readable_description, indent = 3)